gem/interpreter.py

```diff
@@ -29,6 +29,27 @@ def __init__(self, arr, fids=None):
         self.arr = arr
         self.fids = fids if fids is not None else ()
 
+    def broadcast(self, fids):
+        """Given some free indices, return a broadcasted array which
+        contains extra dimensions that correspond to indices in fids
+        that are not in ``self.fids``.
+
+        Note that inserted dimensions will have length one.
+
+        :arg fids: The free indices for broadcasting.
+        """
+        # Select free indices
+        axes = tuple(self.fids.index(fi) for fi in fids if fi in self.fids)
+        assert len(axes) == len(self.fids)
+        # Add shape
+        axes += tuple(range(len(self.fids), self.arr.ndim))
+        # Move axes, insert extra axes
+        arr = numpy.transpose(self.arr, axes)
+        for i, fi in enumerate(fids):
+            if fi not in self.fids:
+                arr = numpy.expand_dims(arr, axis=i)
+        return arr
```

Comment:
`numpy.transpose` in the generic sense is a reordering of axes/dimensions, so yes, we need to do that, because free indices might appear in different order. `numpy.reshape`: sort of, but not quite. No axes are split or merged, but length-one dimensions are inserted, which will expand to the correct length through broadcasting.